[lldb] Enable support for DWARF64 format handling

HemangGadhavi opened the pull request:
This PR introduces support for the DWARF64 format, enabling handling of 64-bit DWARF sections as defined by the DWARF specification. The update includes adjustments to header parsing and modification of form values to accommodate 64-bit offsets and values. 
Also Added the testcase to verify the DWARF64 format.

HemangGadhavi committed:
[lldb][DWARF64] Enable support for DWARF64 format handling

Michael137 reviewed (commented):
Is there a test-plan for DWARF64? Is there a buildbot running the test-suite with DWARF64 somewhere?

DavidSpickett commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp:
```diff
@@ -1073,20 +1073,7 @@ const lldb_private::DWARFDataExtractor &DWARFUnit::GetData() const {
-  }
-  llvm_unreachable("invalid UnitType.");
-}
+uint32_t DWARFUnit::GetHeaderByteSize() const { return m_header.getSize(); }
```
This should be it's own NFC change PR.

HemangGadhavi replied:
okay sure, I will create separate PR for this changes. 
Thanks for suggestion.

DavidSpickett commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFFormValue.cpp:
```diff
@@ -165,26 +172,27 @@ static FormSize g_form_sizes[] = {
-    {1, 4},  // 0x17 DW_FORM_sec_offset
+    {1, 1}, // 0x11 DW_FORM_ref1
+    {1, 2}, // 0x12 DW_FORM_ref2
+    {1, 4}, // 0x13 DW_FORM_ref4
+    {1, 8}, // 0x14 DW_FORM_ref8
+    {0, 0}, // 0x15 DW_FORM_ref_udata
+    {0, 0}, // 0x16 DW_FORM_indirect
+    {0,
+     0}, // 0x17 DW_FORM_sec_offset (4 bytes for DWARF32, 8 bytes for DWARF64)
     {0, 0},  // 0x18 DW_FORM_exprloc
     {1, 0},  // 0x19 DW_FORM_flag_present
     {0, 0},  // 0x1a DW_FORM_strx (ULEB128)
     {0, 0},  // 0x1b DW_FORM_addrx (ULEB128)
     {1, 4},  // 0x1c DW_FORM_ref_sup4
     {0, 0},  // 0x1d DW_FORM_strp_sup (4 bytes for DWARF32, 8 bytes for DWARF64)
     {1, 16}, // 0x1e DW_FORM_data16
-    {1, 4},  // 0x1f DW_FORM_line_strp
-    {1, 8},  // 0x20 DW_FORM_ref_sig8
+    {0, 0}, // 0x1f DW_FORM_line_strp (4 bytes for DWARF32, 8 bytes for DWARF64)
+    {1, 8}, // 0x20 DW_FORM_ref_sig8
```
Stick to the original formatting here so the diff is a bit cleaner. If clang-format complains about that, just ignore it.

HemangGadhavi replied:
Yes exactly clang-format complaining that's why I modified. 
But yes I will change it.

DavidSpickett reviewed (commented):
Also wondering about testing this. What tools can produce DWARF64?

Have you ran `check-lldb` with DWARF64 enabled and what results did you get?

I doubt anyone has spare capacity to add a whole new bot for this, but perhaps we would be ok with a few new tests that force DWARF64. Covering the main differences.

labath commented:
> Clang only supports DWARF64 for ELF:

I'm surprised that is the case, but I don't think it matters because I think you should be able to test all of this stuff without ever running a binary (which means you can build an elf target regardless of the host platform).

That said, I think it would be best to avoid using clang for this sort of stuff, as very hard to tell what it actually tests. Like, you can't tell by looking at the c++ source code whether the resulting dwarf will use (e.g.) DW_FORM_line_strp or not. I'd recommend using assembly or dwarf yaml for that. E.g. you can take one of the simple tests from test/Shell/SymbolFile/DWARF/x86 and modify the assembly so it generates dwarf64.

JDevlieghere commented:
+1 on testing and Pavel's suggestion to use assembly/yaml for it so it can run everywhere.

HemangGadhavi committed:
Added testcase for the DWARF64 and review comments

HemangGadhavi commented:
@labath @DavidSpickett @JDevlieghere @Michael137 
Thanks everyone for the valuable feedback. 

I have addressed all the review comments please verify once again. 

> We should do something similar. We already have the FormParams object accessible from inside our DWARFUnit (through the m_header field), so I think it'd be best to replace `GetFormat` method you added with `GetFormParams`, and then the parsing code would do a `m_unit->GetFormParams().getDwarfOffsetByteSize`

@labath I have added the `GetFormParams()` and remove `if dwarf64()`.

On the Testing front, I have created the unit test case for `DWARF64` format, which basically testing the `.debug_info` header(including `DWARF64` format, address size) and the Compilation Unit header, 
Also its testing the `DWARF64` related attributes which are included in this PR (i.e. DW_FORM_sec_offset & DW_FORM_strp).

Please review once and give your feedback, also let me know is there any additional testing required for the same ?

Please Note that, clang-format is failling but not because of this PR changes but because of some existing format.

HemangGadhavi commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFFormValue.cpp:
```diff
@@ -121,8 +124,12 @@ bool DWARFFormValue::ExtractValue(const DWARFDataExtractor &data,
+          ref_addr_size = 4;
+        else if (m_unit->GetFormat() == DwarfFormat::DWARF64)
+          ref_addr_size = 8;
+      }
```
Done. Addressed

labath commented on lldb/unittests/SymbolFile/DWARF/DWARF64UnitTest.cpp:
```diff
@@ -0,0 +1,96 @@
+  attrs.ExtractFormValueAtIndex(3, form_value); // Validate DW_AT_low_pc
+  ASSERT_EQ(form_value.Unsigned(), 0xdeadbeef);
+  attrs.ExtractFormValueAtIndex(4, form_value); // Validate DW_AT_high_pc
+  ASSERT_EQ(form_value.Unsigned(), 0xdeadbeef);
```
It'd be better to choose distinct values for these.

.. but maybe you can also delete these as DW_AT_low/high_pc does not depend on dwarf size.

HemangGadhavi replied:
Thanks @labath 
I have removed it. If no more comments can you please merge it. I do not have access yet.

HemangGadhavi committed:
Addressed review comment

HemangGadhavi commented:
@DhruvSrivastavaX can you please merge it

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFFormValue.cpp:
```diff
@@ -119,10 +121,7 @@ bool DWARFFormValue::ExtractValue(const DWARFDataExtractor &data,
-        ref_addr_size = m_unit->GetAddressByteSize();
-      else
-        ref_addr_size = 4;
+      ref_addr_size = m_unit->GetFormParams().getRefAddrByteSize();
```
Outside the scope of this PR but seems like we can just make `ref_addr_size` local to the cases that need it. Don't see the point of having it outside the switch

HemangGadhavi replied:
I think it used in multiple switch cases, Thats why they have made it common for all the case instead for each switch case.

Michael137 replied:
It's reset before it's used in each case though.

HemangGadhavi replied:
Okay Thanks

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFFormValue.cpp:
```diff
@@ -288,7 +284,9 @@ bool DWARFFormValue::SkipValue(dw_form_t form,
     case DW_FORM_line_strp:
-      *offset_ptr += 4;
+      assert(unit); // Unit must be valid
+      ref_addr_size = unit->GetFormParams().getDwarfOffsetByteSize();
```
Same comment re. `ref_addr_size`

JDevlieghere commented on lldb/unittests/SymbolFile/DWARF/DWARF64UnitTest.cpp:
```diff
@@ -0,0 +1,86 @@
+//===-- DWARF64UnitTest.cpp--------------------------------------------=---===//
```
```suggestion
//===-- DWARF64UnitTest.cpp------------------------------------------------===//
```

HemangGadhavi committed:
Update lldb/unittests/SymbolFile/DWARF/DWARF64UnitTest.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

DavidSpickett commented:
Assuming there is not a lot more work to be done for DWARF64, this needs a release note - https://github.com/llvm/llvm-project/blob/main/llvm/docs/ReleaseNotes.md#changes-to-lldb.

You can do that in this PR or a follow up.

HemangGadhavi committed:
Updated ReleaseNotes

HemangGadhavi commented:
> Assuming there is not a lot more work to be done for DWARF64, this needs a release note - https://github.com/llvm/llvm-project/blob/main/llvm/docs/ReleaseNotes.md#changes-to-lldb.
> 
> You can do that in this PR or a follow up.

Updated with this PR only, Please review once @DavidSpickett Thanks!

DavidSpickett commented on llvm/docs/ReleaseNotes.md:
````diff
@@ -299,6 +299,7 @@ Changes to LLDB
     stop reason = SIGSEGV: sent by tkill system call (sender pid=649752, uid=2667987)
   ```
 * ELF Cores can now have their siginfo structures inspected using `thread siginfo`.
+* LLDB now supports `DWARF64` debugging information format.
````
I would say "the `DWARF64`".

HemangGadhavi replied:
Thanks! Updated

DavidSpickett commented:
This has approvals already and looks good to me, do you need someone to merge it?

DavidSpickett commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFFormValue.cpp:
```diff
@@ -175,15 +174,16 @@ static FormSize g_form_sizes[] = {
     {0, 0},  // 0x16 DW_FORM_indirect
-    {1, 4},  // 0x17 DW_FORM_sec_offset
+    {0,
+     0}, // 0x17 DW_FORM_sec_offset (4 bytes for DWARF32, 8 bytes for DWARF64)
```
Actually, this line got mangled, put it back on one same line.

HemangGadhavi replied:
It exceeds 80 characters when written on a single line, so to avoid and It auto set by clang-format.

HemangGadhavi committed:
small correction

DavidSpickett commented:
Please go ahead and merge this, or I can do so for you if you're happy with it.
